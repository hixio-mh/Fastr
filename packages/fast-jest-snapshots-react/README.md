# FAST Jest snapshots React

`@microsoft/fast-jest-snapshots-react` is a small utility library for creating snapshot Jest tests for React components.
Given a single component, the `generateSnapshots` function will iterate over a set of prop data and create a snapshot test for each `prop` instance.

## Usage

To use, import the `ISnapshotTestSuite` interface and the `generateSnapshots` function from `@microsoft/fast-jest-snapshots-react`:

```ts
import {generateSnapshots, ISnapshotTestSuite} from "@microsoft/fast-jest-snapshots-react";
```

Then, create you test data snapshot suite, where every index of `data` is a valid prop object for your `component`:

```ts
// in a Jest test file
import {MyReactComponent, IMyReactComponentProps} from "my-react-component.tsx";
import {ISnapshotTestSuite, generateSnapshots} from "@microsoft/fast-jest-snapshots-react";

const testSuite: ISnapshotTestSuite<IMyReactComponentProps> = {
    name: 'test component',
    component: MyCustomReactComponentConstructor,
    data: [
        {
            id: "test-id",
            text: "hello world",
            // other component props
        }
    ]
};

// Generate snapshots using the prop and component data you defined
generateSnapshots(testSuite);
```

### Snapshot descriptions
Descriptions can optionally be applied to any snapshot for providing context to the test and the expectation of the test.
To do this, simply put the `prop` object into an array where the first index is the snapshot description and the second index is the `prop` object:

```ts
// in a Jest test file
import {MyReactComponent, IMyReactComponentProps} from "my-react-component.tsx";
import {ISnapshotTestSuite, generateSnapshots} from "@microsoft/fast-jest-snapshots-react";

const testSuite: ISnapshotTestSuite<IMyReactComponentProps> = {
    // ...
    data: [
        [
            "should render with an id attribute",
            {
                id: "test-id",
                text: "hello world",
                // other component props
            }

        ]
    ]
};
```
