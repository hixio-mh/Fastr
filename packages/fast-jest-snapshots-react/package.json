{
  "name": "@microsoft/fast-jest-snapshots-react",
  "description": "A small utility for automating Jest snapshot creation of React components",
  "sideEffects": false,
  "version": "1.8.0",
  "author": {
    "name": "Microsoft",
    "url": "https://discord.gg/FcSNfg4"
  },
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/Microsoft/fast-dna.git"
  },
  "main": "./dist/index.js",
  "types": "./dist/index.d.ts",
  "scripts": {
    "prepare": "tsc -p ./tsconfig.json",
    "test": "npm run tslint",
    "start": "webpack-dev-server",
    "tslint": "tslint -c ./tslint.json '**/*.ts{,x}'",
    "tslint:fix": "tslint -c ./tslint.json --fix '**/*.ts{,x}'"
  },
  "devDependencies": {
    "@microsoft/fast-tslint-rules": "^1.8.0",
    "@types/jest": "^22.2.2",
    "@types/react": "^16.3.6",
    "jest": "^22.4.3",
    "react": "^16.3.0",
    "ts-jest": "^22.4.2",
    "tslint": "^5.9.1",
    "typescript": "^2.8.1"
  },
  "dependencies": {
    "react-test-renderer": "^16.3.0"
  },
  "peerDependencies": {
    "jest": "^22.0.0",
    "react": "^16.3.0"
  }
}
