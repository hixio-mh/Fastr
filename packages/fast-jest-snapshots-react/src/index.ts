import React from "react";
import * as renderer from "react-test-renderer";

/**
 * A set of snapshot test cases
 */
export type SnapshotTestCases<T> = Array<SnapshotTestCase<T>>;

/**
 * An prop object to supply for snapshots. Optionally, a
 * an array can be provided where the first index is a string representing
 * a snapshot description, with the second index being the prop data
 */
export type SnapshotTestCase<T> = T | [string, T];

/**
 * An interface describing component example objects used for snapshots and component testing.
 */
export interface SnapshotTestSuite<T> {
    /**
     * The name of the component
     */
    name: string;

    /**
     * The component constructor
     */
    component: React.ComponentType<T>;

    /**
     * An array of prop instances for the component
     */
    data: SnapshotTestCases<T>;

    /**
     * The JSON schema for the components data
     * @deprecated - this property is not required for snapshot testing and will be removed
     */
    schema?: any;

    /**
     * The detail view component data
     * @deprecated - this property is not required for snapshot testing and will be removed
     */
    detailData?: T;

    /**
     * Documentation for the component
     * @deprecated - this property is not required for snapshot testing and will be removed
     */
    documentation?: JSX.Element;
}

/**
 * Executes a single snapshot test given a component, component data, and a test title
 */
export function renderSnapshot<T>(
    data: T,
    component: React.ComponentType<T>,
    title: string
): void {
    test(
        title,
        (): void => {
            const renderedComponent: any = renderer.create(
                React.createElement(component, data)
            );
            const componentJson: JSON = renderedComponent.toJSON();

            expect(componentJson).toMatchSnapshot();
        }
    );
}

/**
 * Generate a set of snapshot tests given a snapshot suite
 */
export function generateSnapshots<T>(examples: SnapshotTestSuite<T>): void {
    const data: SnapshotTestCases<T> = examples.data;
    const component: React.ComponentType<T> = examples.component;

    if (Array.isArray(data)) {
        data.forEach(
            (example: SnapshotTestCase<T>, index: number): void => {
                let title: string;
                let props: T;

                if (Array.isArray(example)) {
                    title = `${examples.name} ${example[0]}`;
                    props = example[1];
                } else {
                    title = `${examples.name}: ${index}`;
                    props = example;
                }

                renderSnapshot(props, component, title);
            }
        );
    }
}
