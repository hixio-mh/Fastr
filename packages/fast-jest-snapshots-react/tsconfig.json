{
    "compilerOptions": {
        "moduleResolution": "node",
        "module": "commonjs",
        "target": "ES6",
        "baseUrl": ".",
        "declaration": true,
        "jsx": "react",
        "outDir": "./dist"
    },
    "include": [
        "src/**/*"
    ]
}
