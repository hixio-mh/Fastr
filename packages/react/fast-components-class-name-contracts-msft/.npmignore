# Source files
src/

# Configs
tsconfig.json