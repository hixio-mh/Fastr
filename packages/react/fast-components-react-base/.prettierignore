coverage/*
dist/*
www/*
src/**/.tmp/