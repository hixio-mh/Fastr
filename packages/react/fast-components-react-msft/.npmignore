# Tests
__test__/
*.spec.*
*.test.*

# Source files
app/
www/
build/
coverage/
src/
assets/

# config
babel.config.js
tsconfig.json
webpack.config.js