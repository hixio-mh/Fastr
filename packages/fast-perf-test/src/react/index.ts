import React from "react";
import ReactDOM from "react-dom";
import { Checkbox, TextField } from "@microsoft/fast-components-react-msft";

const root = document.createElement("div");
document.body.appendChild(root);

function getQueryVariable(variable) {
    var query = window.location.search.substring(1);
    var vars = query.split("&");
    for (var i = 0; i < vars.length; i++) {
        var pair = vars[i].split("=");
        if (decodeURIComponent(pair[0]) == variable) {
            return decodeURIComponent(pair[1]);
        }
    }
}

const count = getQueryVariable("count");

function renderItems() {
    return React.createElement(
        React.Fragment,
        {},
        new Array(typeof count === "string" ? (count as any) * 1 : 2000)
            .fill(0)
            .map((item, index) => React.createElement(TextField))
    );
}

function render(): void {
    ReactDOM.render(React.createElement(renderItems), root);
}

render();
