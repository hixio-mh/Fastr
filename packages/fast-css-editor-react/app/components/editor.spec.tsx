import * as React from "react";
import { generateSnapshots } from "@microsoft/fast-jest-snapshots-react";
import { SnapshotTestSuite } from "@microsoft/fast-jest-snapshots-react";
import CSSEditorSchema from "../../src/editor.schema.json";
import CSSEditor, { CSSEditorProps, PositionValue } from "../../src";
import { noop } from "lodash-es";

const name: string = "css-editor";

const exampleData: CSSEditorProps = {
    position: PositionValue.static,
    onChange: noop,
};

const detailData: CSSEditorProps = {
    position: PositionValue.absolute,
    bottom: "8",
    right: "8",
    onChange: noop,
};

const examples: SnapshotTestSuite<CSSEditorProps> = {
    name,
    component: CSSEditor,
    schema: CSSEditorSchema,
    detailData,
    data: [exampleData],
};

describe(name, () => {
    generateSnapshots(examples);
});
