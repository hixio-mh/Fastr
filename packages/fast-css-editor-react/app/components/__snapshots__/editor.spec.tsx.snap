// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`css-editor css-editor: 0 1`] = `
<div>
  <div
    className="cssPosition-0-1-1"
  >
    <span
      className="cssPosition_selectContainer-0-1-2"
    >
      <select
        className="cssPosition_selectContainer_select-0-1-3"
        data-location="position"
        onChange={[Function]}
        value="static"
      >
        <option
          value="static"
        >
          Static
        </option>
        <option
          value="absolute"
        >
          Absolute
        </option>
      </select>
    </span>
  </div>
</div>
`;
